libafl_frida/src/asan_errors.rs

```diff
@@ -0,0 +1,569 @@
+pub(crate) struct AsanReadWriteError {
+    pub registers: [usize; 32],
+    pub pc: usize,
+    pub fault: (u16, u16, usize, usize),
+    pub metadata: AllocationMetadata,
+    pub backtrace: Backtrace,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, SerdeAny)]
+pub(crate) enum AsanError {
+    OobRead(AsanReadWriteError),
+    OobWrite(AsanReadWriteError),
+    ReadAfterFree(AsanReadWriteError),
+    WriteAfterFree(AsanReadWriteError),
+    DoubleFree((usize, AllocationMetadata, Backtrace)),
+    UnallocatedFree((usize, Backtrace)),
+    Unknown(([usize; 32], usize, (u16, u16, usize, usize), Backtrace)),
+    Leak((usize, AllocationMetadata)),
+    StackOobRead(([usize; 32], usize, (u16, u16, usize, usize), Backtrace)),
+    StackOobWrite(([usize; 32], usize, (u16, u16, usize, usize), Backtrace)),
+    BadFuncArgRead((String, usize, usize, Backtrace)),
+    BadFuncArgWrite((String, usize, usize, Backtrace)),
+}
+
+impl AsanError {
+    fn description(&self) -> &str {
+        match self {
+            AsanError::OobRead(_) => "heap out-of-bounds read",
+            AsanError::OobWrite(_) => "heap out-of-bounds write",
+            AsanError::DoubleFree(_) => "double-free",
+            AsanError::UnallocatedFree(_) => "unallocated-free",
+            AsanError::WriteAfterFree(_) => "heap use-after-free write",
+            AsanError::ReadAfterFree(_) => "heap use-after-free read",
+            AsanError::Unknown(_) => "heap unknown",
+            AsanError::Leak(_) => "memory-leak",
+            AsanError::StackOobRead(_) => "stack out-of-bounds read",
+            AsanError::StackOobWrite(_) => "stack out-of-bounds write",
+            AsanError::BadFuncArgRead(_) => "function arg resulting in bad read",
+            AsanError::BadFuncArgWrite(_) => "function arg resulting in bad write",
+        }
+    }
+}
+
+/// A struct holding errors that occurred during frida address sanitizer runs
+#[allow(clippy::unsafe_derive_deserialize)]
+#[derive(Debug, Clone, Serialize, Deserialize, SerdeAny)]
+pub struct AsanErrors {
+    options: FridaOptions,
+    errors: Vec<AsanError>,
+}
+
+impl AsanErrors {
+    /// Creates a new `AsanErrors` struct
+    #[must_use]
+    pub fn new(options: FridaOptions) -> Self {
+        Self {
+            options,
+            errors: Vec::new(),
+        }
+    }
+
+    /// Clears this `AsanErrors` struct
+    pub fn clear(&mut self) {
+        self.errors.clear()
+    }
+
+    /// Gets the amount of `AsanErrors` in this struct
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.errors.len()
+    }
+
+    /// Returns `true` if no errors occurred
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.errors.is_empty()
+    }
+
+    /// Get a mutable reference to the global [`AsanErrors`] object
+    pub fn get_mut<'a>() -> &'a mut Self {
+        unsafe { ASAN_ERRORS.as_mut().unwrap() }
+    }
+
+    #[allow(clippy::too_many_lines)]
+    /// Report an error
```

Comment:
I usually put the comments above clippy but that's personal preference

Reply:
Done.